addons/survey/models/survey_survey.py

```diff
@@ -1270,3 +1260,33 @@ def _handle_certification_badges(self, vals):
             # delete all challenges and goals because not needed anymore (challenge lines are deleted in cascade)
             challenges_to_delete.unlink()
             goals_to_delete.unlink()
+
+    # ------------------------------------------------------------
+    # TOOLING / MISC
+    # ------------------------------------------------------------
+
+    def _generate_session_codes(self, code_count=1, excluded_codes=False):
+        """ Generate {code_count} session codes for surveys.
+
+        We try to generate 4 digits code first and see if we have {code_count} unique ones.
+        Then we raise up to 5 digits if we need more, etc until up to 10 digits. """
+
+        self.flush_model(['session_code'])
+
+        session_codes = set()
+        excluded_codes = excluded_codes or set()
+        existing_codes = self.sudo().search_read(
```

Comment:
Not a fan of pre-fetching all existing codes, but I suppose we would not have millions of surveys.